utils/git/git.go

```diff
@@ -0,0 +1,204 @@
+				if err != nil {
+					return
+				}
+
+				entry, ok := <-allEntries
+				if !ok {
+					return
+				}
+
+				totalEntries.Inc()
+				if totalEntries.Load() > c.cfg.MaxEntries {
+					err = fmt.Errorf("%w: maximum entries count exceeded [%d > %d]", commonerrors.ErrTooLarge, totalEntries.Load(), c.cfg.MaxEntries)
+					return
+				}
+
+				if entry.TreeDepth > c.cfg.MaxTreeDepth {
+					err = fmt.Errorf("%w: maximum tree depth exceeded [%d > %d]", commonerrors.ErrTooLarge, entry.TreeDepth, c.cfg.MaxTreeDepth)
+					return
+				}
+
+				mode := entry.TreeEntry.Mode
+				switch {
+				case mode&0o170000 == 0o40000:
+					// Tree
+					tree, subErr := c.repo.TreeObject(entry.TreeEntry.Hash)
+					if subErr != nil {
+						err = subErr
+						return
+					}
+					for i := range tree.Entries {
+						allEntries <- Entry{
+							TreeEntry: tree.Entries[i],
+							TreeDepth: entry.TreeDepth + 1,
+						}
+					}
+				case mode&0o170000 == 0o160000:
+					// Commit (i.e., submodule)
+				case mode&0o170000 == 0o120000:
+					// Symlink
+				default:
+					// Blob
+					blob, subErr := c.repo.BlobObject(entry.TreeEntry.Hash)
+					if subErr != nil {
+						err = subErr
+						return
+					}
+
+					totalSize.Add(blob.Size)
+					if totalSize.Load() > c.cfg.MaxRepositorySize {
+						err = fmt.Errorf("%w: maximum repository size exceeded [%d > %d]", commonerrors.ErrTooLarge, totalSize.Load(), c.cfg.MaxRepositorySize)
+						return
+					}
+
+					if blob.Size > c.cfg.MaxFileSize {
+						err = fmt.Errorf("%w: maximum individual file size exceeded [%d > %d]", commonerrors.ErrTooLarge, blob.Size, c.cfg.MaxFileSize)
+						return
+					}
+
+					totalFileCount.Inc()
+					if totalFileCount.Load() > c.cfg.MaxFileCount {
+						err = fmt.Errorf("%w: maximum file count exceeded [%d > %d]", commonerrors.ErrTooLarge, totalFileCount.Load(), c.cfg.MaxFileCount)
+						return
+					}
+				}
+			}
+			wg.Done()
+			return nil
+		})
+	}
+	err = errs.Wait()
+	if err != nil {
+		return
+	}
+	wg.Wait()
+	return
+}
+
+func (c *CloneObject) Checkout(opts *git.CheckoutOptions) (err error) {
+	worktree, err := c.repo.Worktree()
+	if err != nil {
+		return
+	}
+	return worktree.Checkout(opts)
+}
+
+func NewCloneObject(cfg RepositoryConfig, path string) (c CloneObject, err error) {
+	err = c.Initialise(path, cfg)
+	return
+}
+
+func CloneWithLimits(ctx context.Context, url, dir string, limits ILimits) (err error) {
+	c, err := NewCloneObject(RepositoryConfig{
+		MaxTreeDepth:      limits.GetMaxTreeDepth(),
+		MaxRepositorySize: limits.GetMaxTotalSize(),
+		MaxFileCount:      limits.GetMaxFileCount(),
+		MaxFileSize:       limits.GetMaxFileSize(),
+		MaxEntries:        limits.GetMaxEntries(),
+		CloneOptions: &git.CloneOptions{
+			URL: url,
+		},
+	}, dir)
+	if err != nil {
+		return
+	}
+
+	err = c.ValidateRepository(ctx)
+	if err != nil {
+		return
+	}
+
+	return c.Checkout(&git.CheckoutOptions{Branch: plumbing.ReferenceName("HEAD")})
```

Comment:
could the checkout logic in https://github.com/Arm-Debug/workspace-service-api/blob/063adb24d1ed34226db9a883a5415173ec0faeb4/api/workspace/git_workspace.go#L77 be moved here so that we can checkout the correct branch/tag/hash and apply the general options (e.g. submodules, depth, ...)